src/cryptonote_core/service_node_list.cpp

```diff
@@ -1812,10 +1801,90 @@ namespace service_nodes
       }
     }
 
-    uint64_t amount_left = staking_requirement - total_reserved;
-    const uint64_t DUST  = MAX_NUMBER_OF_CONTRIBUTORS;
-    if (amount_left <= DUST)
-      result.portions[0] += amount_left;
+    //
+    // FIXME(doyle): FIXME(loki) !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+    // This is temporary code to redistribute the insufficient portion dust
+    // amounts between contributors. It should be removed in HF12.
+    //
+    std::array<uint64_t, MAX_NUMBER_OF_CONTRIBUTORS * service_nodes::MAX_KEY_IMAGES_PER_CONTRIBUTOR> excess_portions;
+    std::array<uint64_t, MAX_NUMBER_OF_CONTRIBUTORS * service_nodes::MAX_KEY_IMAGES_PER_CONTRIBUTOR> min_contributions;
+    {
+      // NOTE: Calculate excess portions from each contributor
+      uint64_t loki_reserved = 0;
+      for (size_t index = 0; index < addr_to_portions.size(); ++index)
+      {
+        addr_to_portion_t const &addr_to_portion = addr_to_portions[index];
+        uint64_t min_contribution_portions       = service_nodes::get_min_node_contribution_in_portions(hf_version, staking_requirement, loki_reserved, index);
+        uint64_t loki_amount                     = service_nodes::portions_to_amount(staking_requirement, addr_to_portion.portions);
+        loki_reserved                           += loki_amount;
+
+        uint64_t excess = 0;
+        if (addr_to_portion.portions > min_contribution_portions)
+          excess = addr_to_portion.portions - min_contribution_portions;
+
+        min_contributions[index] = min_contribution_portions;
+        excess_portions[index]   = excess;
+      }
+    }
+
+    uint64_t portions_left  = STAKING_PORTIONS;
+    uint64_t total_reserved = 0;
+    for (size_t i = 0; i < addr_to_portions.size(); ++i)
+    {
+      addr_to_portion_t &addr_to_portion = addr_to_portions[i];
+      uint64_t min_portions = get_min_node_contribution_in_portions(hf_version, staking_requirement, total_reserved, i);
+
+      if (addr_to_portion.portions < min_portions)
+      {
+          // NOTE: Steal dust portions from other contributor if we fall below
+          // the minimum by a dust amount.
+          uint64_t needed = min_portions - addr_to_portion.portions;
+          const uint64_t DUST = MAX_NUMBER_OF_CONTRIBUTORS;
+          if (needed > DUST)
+            continue;
```

Comment:
I think that the allowable DUST needs to be much, much larger; if you allow for an error of one atomic currency unit, that corresponds to about 683212 portion units (at a 27000 stake).  If I'm following the thread properly here, there is a LOKI -> portions -> LOKI -> portions conversion, so it seems like the dust on one contribution could end up being up to 2 atomic units + 2 portions off.

Perhaps DUST threshold here should be multiplied by some "large" number, say 0x100000 (= 2^24), which would correspond to a 13 atomic LOKI error on a 15k stake; 25 atomic units on a 27k stake.  That's probably still a bit bigger than is needed, but seems small enough that no one would notice or care.

Reply:
Had some offline discussion about this and we'll add a fudge factor just to be extra sure since this is temporary code. My initial impression is that since we're always dividing by 1, 2, 3 or 4 (depending on how many contributions are left) then dust will always in actuality be 1-3 portions so long as you apply the dust at the final step in the conversion process.

But it doesn't hurt to have and to avoid any further oversights in the logic here, there's no harm in setting dust to be `STAKING_PORTIONS/staking_requirement` (how much portions in an atomic loki) with the fudge factor we decided on here is 10,  i.e `STAKING_PORTIONS / staking_requirement * 10` 

Not sure I get the `0x100000` as the multiplier here in this case as that seems abnormally large, unless you meant like `2 portions * 0x100000`

Otherwise then if portions are set and resolved correctly, then when we do the conversion from portions to loki for the last time in the wallet, clientside, we have code that will handle the last of the dust amounts from going portions->loki (contributors * locked key images) of 1-3 atomic Loki units.

Reply:
I thought I had another one that was failing with a larger error, but after checking it again (and properly finding and decoding the varint potions from the extra data—not a fun task—indeed the failing registration was off by just 1 in one of the portions, so the extra safety I suggested above may indeed be unwarranted.

> Not sure I get the 0x100000 as the multiplier here in this case as that seems abnormally large, unless you meant like 2 portions * 0x100000

It was a fairly arbitrary choice, but indeed I meant it as a value denominated in portions.  But as per the above, I'm not sure it's actually needed (especially with #522 applied).